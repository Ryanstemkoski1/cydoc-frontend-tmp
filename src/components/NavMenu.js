import React, {Component} from 'react';
import { Container, Menu, Header } from 'semantic-ui-react';
import PropTypes from 'prop-types';

export default class NavMenu extends Component {
    render() {
        return (
            <Menu secondary borderless style={{height: "10vh", borderColor: "white"}} attached={this.props.attached}>
                <Container>
                    <Menu.Item href="/home">
                        <Header as="h2">
                        cydoc
                        </Header>
                    </Menu.Item>
                    <Menu.Menu position="right">
                        <Menu.Item name="create_note" href="/createnote">
                            Create Note
                        </Menu.Item>
                        <Menu.Item name="about" href="/about">
                            About
                        </Menu.Item>
                        <Menu.Item name="login" href="/login">
                            Login
                        </Menu.Item>
                        <Menu.Item name="register" href="/login">
                            Register
                        </Menu.Item>
                    </Menu.Menu>
                </Container>
            </Menu>
        );
    }
};

NavMenu.propTypes = {
  attached: PropTypes.string
};