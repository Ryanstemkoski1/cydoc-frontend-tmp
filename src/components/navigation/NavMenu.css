
/* Navigation Bar has a thicker drop shadow */
.nav-menu {
    box-shadow: 0 0 20px 5px RGBA(0, 128, 159, 0.2) !important;
    height: 10vh;
}

/* Navigation button styles */
.nav-menu .button {
    background-color: #CDE1E9;
    color: #034353;
    font-weight: 400;
}

.nav-menu-logo img {
    margin-right: 10px;
    height: 8vh !important;
    width: auto !important;
}

.nav-menu-brand img {
    padding-bottom: 0.5vh;
    height: 6vh !important;
    width: auto !important;
}

/* Make the divide between menu tabs seamless */
.edit-note-nav-menu {
    border: 0 !important;
}


/* note name input box styles */
.note-name-menu-item input {
    width: 200px;
    color: #0d5e70 !important;

}
.note-name-menu-item .input:hover {
    border-bottom: 0;
}


.note-name-menu-item {
    padding: 0 !important;
}
