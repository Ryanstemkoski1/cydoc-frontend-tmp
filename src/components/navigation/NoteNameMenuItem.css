@media screen and (min-width: 501px) and (max-width: 700px) {
    div.fitted.item.note-name-menu-item {
        max-width: 350px;
        min-width: 250px;
    }

    div.ui.huge.transparent.input.note-title {
        margin-top: 0px;
    }
}


@media screen and (max-width: 500px) {
    div.fitted.item.note-name-menu-item {
        max-width: 150px;
    }
    
    .input {
        overflow-x: hidden;
    }
    .input.content-dropdown.medication {
        overflow-x: visible;
    }

    div.ui.huge.transparent.input.note-title {
        margin-top: 0px;
    }
}
