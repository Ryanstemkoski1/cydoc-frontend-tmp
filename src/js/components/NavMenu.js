import React, {Component, Fragment} from 'react';
import {Container, Header, Icon, Menu} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import {logout} from "../actions";
import {connect} from "react-redux";
import {Link} from "react-router-dom";

function mapDispatchToProps(dispatch) {
    return {
        logout: () => dispatch(logout())
    };
}

const mapStateToProps = state => {
    return {
        isLoggedIn: state.isLoggedIn,
        user: state.user
    };
};


//Navigation Bar component that will go at the top of most pages
class ConnectedNavMenu extends Component {
    constructor(props) {
        super(props);
        this.handleLogout = this.handleLogout.bind(this);
    }

    handleLogout(){
        this.props.logout();
    }

    render() {

        return (
            <Menu secondary borderless style={{height: "10vh", borderColor: "white"}} attached={this.props.attached}>
                <Container>
                    <Menu.Item href="/home">
                        <Header as="h2">
                            cydoc
                        </Header>
                    </Menu.Item>
                    <Menu.Menu position="right">
                        {/* Menu will have different options depending on whether the user is logged in or not */}
                        {this.props.isLoggedIn ? <LoggedInMenuItems handleLogout={this.handleLogout} name={this.props.user.firstName}/> : <DefaultMenuItems/>}
                    </Menu.Menu>
                </Container>
            </Menu>
        );
    }
    };

const NavMenu = connect(mapStateToProps, mapDispatchToProps)(ConnectedNavMenu);
export default NavMenu;

NavMenu.propTypes = {
    // optional prop for stacking another menu above/below
    attached: PropTypes.string
};

//Functional component for menu items that show when user is not logged in
function DefaultMenuItems() {
    return <Fragment>
        <Menu.Item name="create_note">
            <Link to={"/createnote"}>Create Note</Link>
        </Menu.Item>
        <Menu.Item name="about">
            <Link to={"/about"}>About</Link>
        </Menu.Item>
        <Menu.Item name="login">
            <Link to={"/login"}>Login</Link>
        </Menu.Item>
        <Menu.Item name="register">
            <Link to={"/login"}>Register</Link>
        </Menu.Item>
    </Fragment>;
}

//Functional component for menu items that show when user is logged in
function LoggedInMenuItems(props) {
    return <Fragment>
        <Menu.Item name="create_note">
            <Link to={"/createnote"}>Create Note</Link>
        </Menu.Item>
        <Menu.Item name="about">
            <Link to={"/creategraph"}>Create Template</Link>
        </Menu.Item>
        <Menu.Item name="login">
            <Link to={"/dashboard"}>Load Note</Link>
        </Menu.Item>
        <Menu.Item name="register">
            <Link to={"/dashboard"}>Welcome, {props.name}</Link>
            <Icon name="user" style={{marginLeft: "7px"}}/>
        </Menu.Item>
        <Menu.Item name={"logout"} onClick={props.handleLogout}>
            <Link to={"/login"}>Logout</Link>
        </Menu.Item>
    </Fragment>;
}