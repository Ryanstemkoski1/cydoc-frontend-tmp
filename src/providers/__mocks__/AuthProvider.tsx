/* eslint-disable @typescript-eslint/no-unused-vars */
'use client';

import React, {
    PropsWithChildren,
    createContext,
    useCallback,
    useEffect,
    useMemo,
    useState,
} from 'react';

import { CognitoAuth, CognitoUser, NEW_PASSWORD_REQUIRED } from 'auth/cognito';
import { stringFromError } from '@modules/error-utils';
import { log } from '@modules/logging';
import { UserInfoProvider } from './UserInfoProvider';

export interface AuthContextValues {
    cognitoUser: CognitoUser | null;
    authLoading: boolean;
    loginCorrect: boolean; // username/password correct, still need MFA confirmation
    passwordResetRequired: boolean;
    isSignedIn: boolean; // MFA confirmed
    signIn: (
        email: string,
        password: string
    ) => Promise<{ errorMessage?: string; user?: CognitoUser }>;
    signUp: (
        email: string,
        password: string,
        phoneNumber: string
    ) => Promise<{ errorMessage?: string; user?: CognitoUser }>;
    completeFirstLoginUpdate: (
        password: string,
        phoneNumber: string
    ) => Promise<{ errorMessage?: string; user?: CognitoUser }>;
    verifyMfaCode: (
        code: string
    ) => Promise<{ errorMessage?: string; user?: CognitoUser }>;
    signOut: () => void;
}

export const AuthProviderContext = createContext<AuthContextValues | null>(
    null
);
export const AuthProvider: React.FC<
    Record<string, unknown> & PropsWithChildren<object>
> = ({ children }) => {
    const [cognitoUser, setCognitoUser] = useState<CognitoUser | null>(null);
    const [loginCorrect, setLoginCorrect] = useState(false);
    const [authLoading, setAuthLoading] = useState(true);
    const passwordResetRequired =
        cognitoUser?.challengeName === NEW_PASSWORD_REQUIRED;

    const [isSignedIn, setIsSignedIn] = useState(false);
    // TODO: parse ^^status^^ from cognitoUser when it changes (see below)
    // Here's an async method: https://github.com/aws-amplify/amplify-js/issues/3640#issuecomment-1198905668
    // const isSignedIn = useMemo(() => !!user?.challengeName?.length, [user]); // This key is not the correct way to determine this

    const signOut = useCallback(
        () => {
            return CognitoAuth.signOut()
                .then(() => {
                    setLoginCorrect(false);
                    setIsSignedIn(false);
                    setCognitoUser(null);
                })
                .catch((reason) =>
                    log(`Sign out error: ${stringFromError(reason)}`, {
                        isSignedIn,
                        cognitoUser,
                        loginCorrect,
                        reason,
                    })
                );
        },
        /* don't need to include dependencies that are just for logging */
        // eslint-disable-next-line react-hooks/exhaustive-deps
        []
    );

    // Persist user login status on refresh
    useEffect(() => {
        setAuthLoading(false);
    }, []);

    const signIn: AuthContextValues['signIn'] = useCallback(
        async (email: string, password: string) => {
            console.log('signIn', {
                email,
                password,
            });
            return { user: undefined, errorMessage: 'mock provider' };
        },
        []
    );

    const signUp: AuthContextValues['signUp'] = useCallback(
        async (email: string, password: string, phoneNumber: string) => {
            return { user: undefined, errorMessage: 'mock provider' };
        },
        []
    );

    const verifyMfaCode = useCallback(async (code: string) => {
        return { user: undefined, errorMessage: 'mock provider' };
    }, []);

    const completeFirstLoginUpdate: AuthContextValues['completeFirstLoginUpdate'] =
        useCallback(
            async (password: string, phoneNumber: string) => {
                return {
                    user: undefined,
                    errorMessage: undefined,
                };
            },
            [cognitoUser]
        );

    const contextValue: AuthContextValues = useMemo(() => {
        return {
            authLoading,
            cognitoUser,
            completeFirstLoginUpdate,
            isSignedIn,
            loginCorrect,
            passwordResetRequired,
            signIn,
            signOut,
            signUp,
            verifyMfaCode,
        };
    }, [
        authLoading,
        cognitoUser,
        completeFirstLoginUpdate,
        isSignedIn,
        loginCorrect,
        passwordResetRequired,
        signIn,
        signOut,
        signUp,
        verifyMfaCode,
    ]);

    return (
        <AuthProviderContext.Provider value={contextValue}>
            <UserInfoProvider>{children}</UserInfoProvider>
        </AuthProviderContext.Provider>
    );
};

const waitForCode = () => new Promise((resolve) => setTimeout(resolve, 2500));
