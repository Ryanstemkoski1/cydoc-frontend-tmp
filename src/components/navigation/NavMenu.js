import React, { Component, Fragment } from 'react';
import {Dropdown, Menu, Image, Icon, Button, Input, Container} from 'semantic-ui-react';
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types';

import { DEFAULT_NAV_MENU_MOBILE_BP, LOGGEDIN_NAV_MENU_MOBILE_BP } from "constants/breakpoints.js";
import HPIContext from 'contexts/HPIContext.js';
import AuthContext from "../../contexts/AuthContext";
import LogoLight from '../../assets/logo-light.png'
import LogoName from '../../assets/logo-name.png'
import './NavMenu.css'

// Navigation Bar component that will go at the top of most pages
class ConnectedNavMenu extends Component {

    static contextType = AuthContext;

    constructor(props) {
        super(props);
        this.state = {
            windowWidth: 0,
            windowHeight: 0,
        };
        this.updateDimensions = this.updateDimensions.bind(this);
    }

    componentDidMount() {
        this.updateDimensions();
        window.addEventListener("resize", this.updateDimensions);
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.updateDimensions);
    }

    updateDimensions() {
        let windowWidth = typeof window !== "undefined" ? window.innerWidth : 0;
        let windowHeight = typeof window !== "undefined" ? window.innerHeight : 0;

        this.setState({ windowWidth, windowHeight });
    }

    render() {
        const { windowWidth } = this.state;
        const collapseDefaultNav = windowWidth < DEFAULT_NAV_MENU_MOBILE_BP;
        const collapseLoggedInNav = windowWidth < LOGGEDIN_NAV_MENU_MOBILE_BP;
        return (
            <div>
                <Menu className={this.props.className + " nav-menu"} attached={this.props.attached}>

                    {/* Logo Image - display icon when tabs not collapsed, and brand name when collpased. */}
                    {/* Only logo image is displayed when the note name must be shown next to it */}
                    <Menu.Item>
                        {this.props.displayNoteName && collapseLoggedInNav?
                            null :
                            <Image className="nav-menu-logo" href='/home' src={LogoLight} />
                        }
                        {this.props.displayNoteName ?
                            null :
                            <Image className="nav-menu-brand" size='small' href='/home' src={LogoName} />
                        }
                    </Menu.Item>

                    {/* When parent is EditNote, then display the note name item */}
                    {this.props.displayNoteName ?  <NoteNameMenuItem /> : null}

                    {/* Navigation links */}
                    <Menu.Menu position="right">
                        {/* Menu will have different options depending on whether the user is logged in or not */}
                        {this.context.token ?
                            <LoggedInMenuItems
                                handleLogout={this.context.logOut}
                                name={this.context.user.firstName}
                                collapseNav={collapseLoggedInNav} />
                            :
                            <DefaultMenuItems collapseNav={collapseDefaultNav} />}
                    </Menu.Menu>
                </Menu>
            </div>

        );
    }
};

const NavMenu = ConnectedNavMenu;
export default NavMenu;

NavMenu.propTypes = {
    // optional prop for stacking another menu above/below
    attached: PropTypes.string,
    // optional prop for whether to display or hide hte note name menu item
    displayNoteName: PropTypes.bool
};


// class component that displays and changes note name
// shown only if parent is EditNote.
class NoteNameMenuItem extends Component {

    static contextType = HPIContext

    handleInputChange = (event) => {
        this.setState({textInput: event.target.value})
        this.context.onContextChange("title", event.target.value)
    }

    render () {
        return (
            <Menu.Item className="note-name-menu-item">
                <HPIContext.Consumer>
                    {value =>
                        <>
                            <Input
                                size="huge"
                                transparent
                                placeholder="Untitled Note"
                                onChange={this.handleInputChange}
                                onFocus={()=>{
                                    if (this.context.title === "Untitled Note") {
                                        this.context.onContextChange("title", "")
                                    }
                                }}
                                onBlur={()=>{
                                    this.setState({isTitleFocused: false})
                                    if (this.context.title === '') {
                                        this.context.onContextChange("title", "Untitled Note")
                                    }
                                }}
                                value={this.context.title}
                            />
                            <Button basic onClick={this.context.saveNote} className="save-button">
                                Save
                            </Button>
                        </>
                    }

                </HPIContext.Consumer>
            </Menu.Item>
        )
    }

}

// Functional component for menu items that show when user is not logged in
function DefaultMenuItems(props) {
    return props.collapseNav ?
        (<Menu.Item>
                <Dropdown icon="large bars" className='collapsed-dropdown'>
                    <Dropdown.Menu>
                        <Dropdown.Item as={Link} name="about" to="/about" text="About" />
                        <Dropdown.Item as={Link} name="login" to="/login" text="Login" />
                        <Dropdown.Item as={Link} name="register" to="/register" text="Sign Up" />
                    </Dropdown.Menu>
                </Dropdown>
            </Menu.Item>
        ) : (
            <Fragment>
                <Menu.Item as={Link} name="about" to="/about" text="About" />
                <Menu.Item>
                    <Button.Group>
                        <Button as={Link} basic color={"teal"} name="login" to="/login" text="Login" >Login</Button>
                        <Button as={Link} color={"teal"} name="register" to="/register" text="Sign Up" >Sign Up</Button>
                    </Button.Group>
                </Menu.Item>

            </Fragment>
        );
}

//Functional component for menu items that show when user is logged in
function LoggedInMenuItems(props) {
    return (props.collapseNav ? (
            // (<Menu.Item>
            //         <Dropdown icon="large bars">
            //             <Dropdown.Menu>
            //                 <Dropdown.Item name="welcome" style={{color: '#6DA3B1', fontStyle: 'italic', fontWeight: 'light'}}>
            //                     {props.name}
            //                 </Dropdown.Item>
            //                 <Dropdown.Item as={Link} name="myNotes" to="/dashboard" text="My Notes" />
            //
            //                 <Dropdown.Item name="logout" href="/home" text="Logout" onClick={props.handleLogout} />
            //             </Dropdown.Menu>
            //         </Dropdown>
            //     </Menu.Item>
            <>
                <Menu.Item>

                    <Button.Group>
                        <Button as={Link} name="myNotes" to="/dashboard" text="My Notes" >
                            <Icon name="sticky note outline" className='collapsed-icon'/>
                        </Button>
                        <Button name="logout" href="/home" text="Logout" onClick={props.handleLogout} className='logout-button'>
                            <Icon name="sign out alternate" className='collapsed-icon'/>
                        </Button>
                    </Button.Group>
                </Menu.Item>
            </>
        ) : (
            <>
                <Menu.Item>
                    <Button.Group>
                        <Button as={Link} name="myNotes" to="/dashboard" text="My Notes" >My Notes</Button>
                        <Button name="logout" href="/home" text="Logout" onClick={props.handleLogout} style={{color: '#FC4F56'}}>Log Out</Button>
                    </Button.Group>
                </Menu.Item>
                <Menu.Item name="welcome" style={{color: '#6DA3B1', fontWeight: 'normal'}}>
                    <Icon name="user outline" /> {props.name}
                </Menu.Item>
            </>
        )
    );
}

