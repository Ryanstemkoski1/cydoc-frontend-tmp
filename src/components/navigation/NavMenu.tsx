import React, { useState, useEffect } from 'react';
import { Dropdown, Menu, Header, Image, Icon, Button } from 'semantic-ui-react';
import { Link, useHistory } from 'react-router-dom';
import {
    HIDE_CYDOC_IN_NAV_MENU_BP,
    LOGGEDIN_NAV_MENU_MOBILE_BP,
} from 'constants/breakpoints.js';
import Logo from '../../assets/cydoc-logo.svg';
import NoteNameMenuItem from './NoteNameMenuItem';
import 'pages/EditNote/content/hpi/knowledgegraph/src/css/Button.css';
import './NavMenu.css';
import {
    selectDoctorViewState,
    selectInitialPatientSurvey,
    selectPatientViewState,
} from 'redux/selectors/userViewSelectors';
import { changeUserView, UserViewAction } from 'redux/actions/userViewActions';
import { connect } from 'react-redux';
import { CurrentNoteState } from 'redux/reducers';
import {
    updateActiveItem,
    UpdateActiveItemAction,
} from 'redux/actions/activeItemActions';
import { YesNoResponse } from 'constants/enums';
import { initialSurveyProps } from 'pages/EditNote/content/patientview/InitialSurvey';
import { additionalSurvey } from 'redux/reducers/additionalSurveyReducer';
import { selectActiveItem } from 'redux/selectors/activeItemSelectors';
import constants from '../../constants/constants.json';
import './NavMenu.css';
import useAuth from 'hooks/useAuth';
import useUser from 'hooks/useUser';

interface ConnectedNavMenuProps {
    className?: string;
    // For whether to stack another menu above/below
    attached?: 'top' | 'bottom';
    // Whether to display or hide the note name
    displayNoteName?: boolean;
}

// Navigation Bar component that will go at the top of most pages
const ConnectedNavMenu: React.FunctionComponent<Props> = (props: Props) => {
    const {
        className,
        attached = 'top',
        displayNoteName = false,
        patientView,
        doctorView,
        changeUserView,
        updateActiveItem,
    } = props;

    const { isSignedIn, signOut, loginCorrect } = useAuth();
    const { user, isManager } = useUser();

    // email/password correct but waiting on MFA? allow users to logOut
    const userCurrentlyLoggingIn = loginCorrect && !isSignedIn;

    console.log(`user manager: ${isManager}`, {
        user,
        isManager,
    });

    const [windowWidth, setWindowWidth] = useState(0);

    // Set event listeners for window resize to determine mobile vs web view
    useEffect(() => {
        const updateDimensions = (): void => {
            setWindowWidth(
                typeof window !== 'undefined' ? window.innerWidth : 0
            );
        };
        updateDimensions();
        window.addEventListener('resize', updateDimensions);
        return (): void =>
            window.removeEventListener('resize', updateDimensions);
    }, []);

    // Display warning when user tries to close the tab
    useEffect(() => {
        const handleTabClose = (event: BeforeUnloadEvent) => {
            event.preventDefault();
            // Some browsers do not support overriding this message (ex: Chrome), so the custom message will not display
            // in all browsers.
            return (event.returnValue =
                'Are you sure you want to exit? Your current note will not be saved.');
        };

        window.addEventListener('beforeunload', handleTabClose);

        return () => {
            window.removeEventListener('beforeunload', handleTabClose);
        };
    }, []);

    const collapseLoggedInNav = windowWidth < LOGGEDIN_NAV_MENU_MOBILE_BP;
    const hideCydoc = windowWidth < HIDE_CYDOC_IN_NAV_MENU_BP;

    const dropdownOptions = [
        {
            as: Link,
            to: '/editprofile',
            key: 'editProfile',
            text: 'Edit Profile',
            icon: 'setting',
            selected: false,
            active: window.location.href.includes('editprofile'),
        },
        {
            as: Link,
            to: '/profilesecurity',
            key: 'profileSecurity',
            text: 'Profile Security',
            icon: 'lock',
            selected: false,
            active: window.location.href.includes('profilesecurity'),
        },
        {
            as: Link,
            to: '/',
            key: 'logout',
            text: 'Log Out',
            icon: 'sign out',
            onClick: signOut,
            selected: false,
            active: false,
        },
    ];

    // Use for redirecting
    const history = useHistory();

    const navigateToHome = () => {
        history.push('/dashboard');
    };

    const logoNotLoggedIn = () => {
        history.push('/');
    };

    const loginButton = (
        <Button
            as={Link}
            color='teal'
            name='login'
            to='/login'
            content='Login'
            id='nav-menu__login-button'
        />
    );

    const logOutButton = (
        <Button
            as={Link}
            color='teal'
            name='logOut'
            onClick={signOut}
            content='Log Out'
            id='nav-menu__login-button'
        />
    );

    const userManager = isManager ? (
        <Menu.Item>
            <Button
                basic
                color='teal'
                name='users'
                content={collapseLoggedInNav ? undefined : 'Manager Users'}
                icon='users'
                onClick={() => history.push('/managerdashboard')}
            />
        </Menu.Item>
    ) : null;
    // Menu items when not logged in
    const defaultMenuItems = (
        // email/password correct but waiting on MFA? allow users to logOut
        <Menu.Item>
            {userCurrentlyLoggingIn ? logOutButton : loginButton}
            <Button
                icon='plus'
                content='Sign Up'
                size='small'
                onClick={() => {
                    if (userCurrentlyLoggingIn) {
                        // if the user tries to signUp while waiting for MFA, sign them out
                        signOut();
                    }
                    history.push('/sign-up');
                }}
            />
        </Menu.Item>
    );
    // Menu items when logged in
    const loggedInMenuItems = (
        <>
            {userManager}
            <Menu.Item>
                <Button
                    basic
                    color='teal'
                    name='home'
                    content={collapseLoggedInNav ? undefined : 'Home'}
                    icon='hospital outline'
                    onClick={navigateToHome}
                />
            </Menu.Item>
            <Menu.Item>
                <Dropdown
                    button
                    basic
                    color='teal'
                    floating
                    name='profile'
                    className={`profile-button ${
                        collapseLoggedInNav ? 'profile-mobile' : ''
                    }`}
                    options={dropdownOptions}
                    trigger={
                        <span>
                            <Icon
                                name='user outline'
                                className={
                                    collapseLoggedInNav
                                        ? 'profile-mobile-icon'
                                        : ''
                                }
                            />
                            {collapseLoggedInNav ? undefined : user?.firstName}
                        </span>
                    }
                />
            </Menu.Item>
        </>
    );

    const checkPatientView = () => {
        if (!constants.PATIENT_VIEW_TAB_NAMES.includes(props.activeItem)) {
            return false;
        }
        const { userSurveyState, additionalSurvey } = props;
        const selected =
            userSurveyState.nodes['2'].response === YesNoResponse.Yes ||
            userSurveyState.nodes['3'].response === YesNoResponse.Yes ||
            userSurveyState.nodes['4'].response === YesNoResponse.Yes;
        if (
            !selected ||
            !userSurveyState.nodes['8'].response ||
            !additionalSurvey.legalFirstName ||
            !additionalSurvey.legalLastName ||
            !additionalSurvey.dateOfBirth ||
            !additionalSurvey.socialSecurityNumber
        ) {
            return false;
        }
        return true;
    };

    return (
        <Menu
            className={`${className} nav-menu nav-bar ${
                displayNoteName && doctorView ? 'nav-bar-input' : ''
            }`}
            attached={attached}
        >
            {isSignedIn ? (
                <Menu.Item className='logo-menu' onClick={navigateToHome}>
                    <Image
                        src={Logo}
                        className={
                            collapseLoggedInNav
                                ? 'logo-circle-mobile'
                                : 'logo-circle'
                        }
                    />
                    {!displayNoteName && !hideCydoc && (
                        <Header
                            as='h1'
                            className={`${
                                collapseLoggedInNav
                                    ? 'logo-text-mobile'
                                    : 'logo-text'
                            }`}
                            content='Cydoc'
                        />
                    )}
                </Menu.Item>
            ) : (
                <Menu.Item className='logo-menu' onClick={logoNotLoggedIn}>
                    <Image
                        src={Logo}
                        className={`${
                            collapseLoggedInNav
                                ? 'logo-circle-mobile'
                                : 'logo-circle'
                        }`}
                    />
                    {!displayNoteName && !hideCydoc && (
                        <Header
                            as='h1'
                            className={`${
                                collapseLoggedInNav
                                    ? 'logo-text-mobile'
                                    : 'logo-text'
                            }`}
                            content='Cydoc'
                        />
                    )}
                </Menu.Item>
            )}

            {/* When parent is EditNote, then display the note name item */}
            {displayNoteName && doctorView && (
                <NoteNameMenuItem mobile={collapseLoggedInNav} />
            )}
            {isSignedIn && history.location.pathname.length > 1 ? (
                collapseLoggedInNav ? (
                    <Button.Group>
                        <Button
                            compact
                            onClick={() => {
                                changeUserView('Patient View');
                                if (!checkPatientView()) updateActiveItem('CC');
                            }}
                            className={`hpi-ph-button${
                                patientView ? '-selected' : ''
                            }`}
                        >
                            Patient
                        </Button>
                        <Button.Or />
                        <Button
                            compact
                            onClick={() => changeUserView('Doctor View')}
                            className={`hpi-ph-button${
                                !patientView ? '-selected' : ''
                            }`}
                        >
                            Doctor
                        </Button>
                    </Button.Group>
                ) : (
                    <Button.Group>
                        <Button
                            style={{ maxHeight: '75%' }}
                            onClick={() => {
                                changeUserView('Patient View');
                                if (!checkPatientView()) updateActiveItem('CC');
                            }}
                            className={`hpi-ph-button${
                                patientView ? '-selected' : ''
                            }`}
                        >
                            Patient View
                        </Button>
                        <Button.Or />
                        <Button
                            style={{ maxHeight: '75%' }}
                            onClick={() => changeUserView('Doctor View')}
                            className={`hpi-ph-button${
                                !patientView ? '-selected' : ''
                            }`}
                        >
                            Doctor View
                        </Button>
                    </Button.Group>
                )
            ) : (
                ''
            )}

            {/* Navigation links */}
            <Menu.Menu position='right'>
                {/* Menu will have different options depending on whether the user is logged in or not */}
                {isSignedIn ? loggedInMenuItems : defaultMenuItems}
            </Menu.Menu>
        </Menu>
    );
};

interface DispatchProps {
    changeUserView: (userView: string) => UserViewAction;
    updateActiveItem: (updatedItem: string) => UpdateActiveItemAction;
}

const mapDispatchToProps = {
    changeUserView,
    updateActiveItem,
};

export interface userViewProps {
    patientView: boolean;
    doctorView: boolean;
}
export interface AdditionalSurveyProps {
    additionalSurvey: additionalSurvey;
}

export interface ActiveItemProps {
    activeItem: string;
}

const mapStateToProps = (
    state: CurrentNoteState
): userViewProps &
    initialSurveyProps &
    AdditionalSurveyProps &
    ActiveItemProps => {
    return {
        patientView: selectPatientViewState(state),
        doctorView: selectDoctorViewState(state),
        userSurveyState: selectInitialPatientSurvey(state),
        additionalSurvey: state.additionalSurvey,
        activeItem: selectActiveItem(state),
    };
};

type Props = ConnectedNavMenuProps &
    userViewProps &
    DispatchProps &
    initialSurveyProps &
    AdditionalSurveyProps &
    ActiveItemProps;

export default connect(mapStateToProps, mapDispatchToProps)(ConnectedNavMenu);
