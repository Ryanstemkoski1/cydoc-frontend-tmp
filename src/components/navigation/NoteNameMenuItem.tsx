import React from 'react';
import { connect } from 'react-redux';
import {
    UpdateNoteTitleAction,
    updateNoteTitle,
} from 'redux/actions/currentNoteActions';
import { CurrentNoteState } from 'redux/reducers';
import { initialNoteTitle } from 'redux/reducers/currentNoteReducer';
import { selectNoteTitle } from 'redux/selectors/currentNoteSelectors';
import { Input, InputOnChangeData, Menu } from 'semantic-ui-react';
import './NoteNameMenuItem.css';

interface StateProps {
    note: CurrentNoteState;
    title: string;
}

interface DispatchProps {
    updateNoteTitle: (title: string) => UpdateNoteTitleAction;
}

type NoteNameMenuItemProps = StateProps & DispatchProps;
type FormChangeHandler = (
    e: React.ChangeEvent,
    data: InputOnChangeData
) => void;

interface User {
    username: string;
    [key: string]: string;
}

export interface Context {
    user: User | null;
    token: string | null;
    storeLoginInfo: (user: string, token: string) => void;
    logOut: () => void;
}

// TODO: Connect to backend
// Component that displays and changes note name shown only if parent is EditNote.
const NoteNameMenuItem: React.FunctionComponent<NoteNameMenuItemProps> = (
    props: NoteNameMenuItemProps
) => {
    const { title, updateNoteTitle } = props;

    const formatOnChange = function (
        action: (val: any) => void
    ): FormChangeHandler {
        return (_e, { value }) => action(value);
    };

    return (
        <Menu.Item className='note-name-menu-item' fitted>
            <Input
                aria-label='Note-Title'
                className='note-title'
                transparent
                size='huge'
                placeholder={initialNoteTitle}
                onChange={formatOnChange(updateNoteTitle)}
                onFocus={() =>
                    title === initialNoteTitle && updateNoteTitle('')
                }
                onBlur={() => title === '' && updateNoteTitle(initialNoteTitle)}
                value={title}
            />
        </Menu.Item>
    );
};

export default connect(
    (state: CurrentNoteState) => ({
        note: state,
        title: selectNoteTitle(state),
    }),
    { updateNoteTitle }
)(NoteNameMenuItem);
