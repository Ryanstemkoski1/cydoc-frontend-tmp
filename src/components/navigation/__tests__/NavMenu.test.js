import React from 'react';
import Enzyme, { mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import NavMenu from '../NavMenu';
import NoteNameMenuItem from '../NoteNameMenuItem';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import { AuthProvider } from 'providers/AuthProvider';

Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureStore([]);

const initialState = {
    patientView: true,
    doctorView: false,
    userSurvey: {},
};

const connectStore = (props, state = initialState) => {
    const store = mockStore({ userView: state });
    props = {
        className: '',
        attached: 'top',
        displayNoteName: false,
        ...props,
    };
    return {
        store,
        wrapper: mount(
            <Router>
                <Provider store={store}>
                    <AuthProvider>
                        <NavMenu {...props} />
                    </AuthProvider>
                </Provider>
            </Router>
        ),
    };
};

describe('NavMenu', () => {
    it('renders without crashing', () => {
        const { wrapper } = connectStore();
        expect(wrapper).toBeTruthy();
    });
    it('matches snapshot', () => {
        const { wrapper } = connectStore();
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('does not render note name outside of editnote', () => {
        const { wrapper } = connectStore();
        expect(wrapper.find(NoteNameMenuItem)).toHaveLength(0);
    });
    it('does not render note name when directed if in patient view', () => {
        const { wrapper } = connectStore({ displayNoteName: true });
        expect(wrapper.find(NoteNameMenuItem)).toHaveLength(0);
    });

    it('renders note name when directed if in doctor view', () => {
        const { wrapper } = connectStore(
            { displayNoteName: true },
            { ...initialState, doctorView: true }
        );
        expect(wrapper.find(NoteNameMenuItem)).toHaveLength(1);
    });
});
