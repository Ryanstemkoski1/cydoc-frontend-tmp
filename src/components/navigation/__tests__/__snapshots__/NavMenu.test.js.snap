// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu matches snapshot 1`] = `"<div><div class=\\"ui top attached  nav-menu menu\\"><a class=\\"item logo-menu\\"><img src=\\"cydoc-logo.svg\\" class=\\"ui image logo-circle\\"><h1 class=\\"ui header logo-text\\">Cydoc</h1></a><div class=\\"right menu\\"><div class=\\"item\\"><a name=\\"login\\" id=\\"nav-menu__login-button\\" class=\\"ui teal button\\" role=\\"button\\" href=\\"/login\\">Login</a><button class=\\"ui small button\\"><i aria-hidden=\\"true\\" class=\\"plus icon\\"></i>Sign Up</button></div></div></div></div>"`;
